src/JSInterop/Microsoft.JSInterop/src/DotNetDispatcher.cs

```diff
@@ -133,54 +146,96 @@ private static object InvokeSynchronously(string assemblyName, string methodIden
+                return Array.Empty<object>();
             }
 
-            // Second, convert each supplied value to the type expected by the method
-            var runtime = (JSRuntimeBase)JSRuntime.Current;
-            var serializerStrategy = runtime.ArgSerializerStrategy;
-            for (var i = 0; i < suppliedArgsLength; i++)
+            // There's no direct way to say we want to deserialize as an array with heterogenous
+            // entry types (e.g., [string, int, bool]), so we need to deserialize in two phases.
+            var jsonDocument = JsonDocument.Parse(argsJson);
+            var shouldDisposeJsonDocument = true;
+            try
             {
-                if (parameterTypes[i] == typeof(JSAsyncCallResult))
+                var suppliedArgsLength = jsonDocument.RootElement.GetArrayLength();
+
+                if (suppliedArgsLength != parameterTypes.Length)
                 {
-                    // For JS async call results, we have to defer the deserialization until
-                    // later when we know what type it's meant to be deserialized as
-                    suppliedArgs[i] = new JSAsyncCallResult(suppliedArgs[i]);
+                    throw new ArgumentException($"In call to '{methodIdentifier}', expected {parameterTypes.Length} parameters but received {suppliedArgsLength}.");
                 }
-                else
+
+                // Second, convert each supplied value to the type expected by the method
+                var suppliedArgs = new object[parameterTypes.Length];
+                var index = 0;
+                foreach (var item in jsonDocument.RootElement.EnumerateArray())
                 {
-                    suppliedArgs[i] = serializerStrategy.DeserializeObject(
-                        suppliedArgs[i], parameterTypes[i]);
+                    var parameterType = parameterTypes[index];
+
+                    if (parameterType == typeof(JSAsyncCallResult))
+                    {
+                        // We will pass the JsonDocument instance to JAsyncCallResult and make JSRuntimeBase
+                        // responsible for disposing it.
+                        shouldDisposeJsonDocument = false;
+                        // For JS async call results, we have to defer the deserialization until
+                        // later when we know what type it's meant to be deserialized as
+                        suppliedArgs[index] = new JSAsyncCallResult(jsonDocument, item);
+                    }
+                    else if (IsIncorrectDotNetObjectRefUse(item, parameterType))
+                    {
+                        throw new InvalidOperationException($"In call to '{methodIdentifier}', parameter of type {parameterType.Name} at index {(index + 1)} must be wrapped in a {nameof(DotNetObjectRef<object>)} instance.");
```

Comment:
I suggest we clarify that it's the method signature that's at fault, not the parameter value being supplied.

Also, shouldn't we be saying `DotNetObjectRef<SomeType>`, not `DotNetObjectRef<object>`?

```suggestion
                        throw new InvalidOperationException($"In call to '{methodIdentifier}', parameter of type '{parameterType.Name}' at index {(index + 1)} must be declared as type 'DotNetObjectRef<{parameterType.Name}>' to receive the incoming value.");
```

Please let me know if I'm misunderstanding this. We don't really want to force people to use `DotNetObjectRef<object>` in all cases if that's avoidable.

Reply:
> Also, shouldn't we be saying `DotNetObjectRef<SomeType>`, not `DotNetObjectRef<object>`?

`nameof` ignores generic parameters, so you baically see "must be wrapped in 'DotnetObjectRef'`. Switching to your recommendation would produce a much nicer error.